mmengine/runner/deepspeed_runner.py

```diff
@@ -0,0 +1,771 @@
+        }
+        # save optimizer state dict to checkpoint
+        if save_optimizer:
+            if not self.model.zero_optimization():
+                checkpoint['optimizer'] = self.optim_wrapper.state_dict()
+            else:
+                self.consolidate_state_dict(self.optim_wrapper.state_dict())
+                # Only the main process needs to load the optimizer's state.
+                optim_state = self.get_zero_state_dict()
+                checkpoint['optimizer'] = optim_state
+
+        # model state is stored after pulling optimizer state to handle ZeRO 3.
+        checkpoint['state_dict'] = weights_to_cpu(self.get_state_dict(model))
+
+        if self.model.zero_optimization_partition_weights():
+            self.model.optimizer.checkpoint_event_epilogue()
+
+        # save param scheduler state dict
+        if save_param_scheduler and self.param_schedulers is None:
+            print_log(
+                '`save_param_scheduler` is True but `self.param_schedulers` '
+                'is None, so skip saving parameter schedulers',
+                logger='current',
+                level=logging.WARNING)
+            save_param_scheduler = False
+        if save_param_scheduler:
+            if isinstance(self.param_schedulers, dict):
+                checkpoint['param_schedulers'] = dict()
+                for name, schedulers in self.param_schedulers.items():
+                    checkpoint['param_schedulers'][name] = []
+                    for scheduler in schedulers:
+                        state_dict = scheduler.state_dict()
+                        checkpoint['param_schedulers'][name].append(state_dict)
+            else:
+                checkpoint['param_schedulers'] = []
+                for scheduler in self.param_schedulers:  # type: ignore
+                    state_dict = scheduler.state_dict()  # type: ignore
+                    checkpoint['param_schedulers'].append(state_dict)
+
+        self.call_hook('before_save_checkpoint', checkpoint=checkpoint)
+        save_checkpoint(checkpoint, filepath)
+
+    def consolidate_state_dict(self,
+                               state_dict: Dict[str, Any],
+                               to: int = 0) -> None:
+        r"""
+        Consolidate a list of ``state_dict`` s (one per rank) on the target
+        rank.
+        Arguments:
+            to (int): the rank that receives the optimizer states (default: 0).
+        Raises:
+            RuntimeError: if ``overlap_with_ddp=True`` and this method is
+                called before this :class:`ZeroRedundancyOptimizer` instance
+                has been fully initialized, which happens once
+                :class:`DistributedDataParallel` gradient buckets have been
+                rebuilt.
+        .. warning:: This needs to be called on all ranks.
+        """
+        from torch.distributed.optim.zero_redundancy_optimizer import (
+            _broadcast_object, _recursive_copy_to_device)
```

Comment:
This logic is borrowed from https://github.com/pytorch/pytorch/blob/7b0650d5cf4897089f32c011504d2b2d185cc60a/torch/distributed/optim/zero_redundancy_optimizer.py#L489

Reply:
I'll try to check it!

Reply:
https://github.com/microsoft/DeepSpeed/blob/ee39187d8f07f7efc615f64affa8cc60ccf41eb5/deepspeed/runtime/engine.py#L3285
Maybe this one?

Reply:
I checked the above link. It provides similar functions but has limitations.

1. It only supports ZeRO3
2. It only supports model state.

I think this `consolidate` logic is general and can be used for other similar purposes.
How about adding it in `mmengine/dist/utils`?